File: provider/amplitude/provider.go
package amplitude

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"

	"github.com/dejanradmanovic/event-spec/hooks"
	"github.com/dejanradmanovic/event-spec/provider"
)

// Provider sends analytics events to Amplitude via the batch HTTP API.
//
// Capabilities: Track ✅, Identify ✅, Group ✅, Page ❌, Alias ✅.
// Page returns ErrUnsupportedOperation — Amplitude has no native page concept.
type Provider struct {
	apiKey      string
	endpoint    string
	transport   *provider.Transport
	queue       *provider.Queue
	rateLimiter *provider.RateLimiter
}

// New constructs and starts an Amplitude Provider from cfg.
// The API key is resolved according to cfg.SecretType before the provider starts.
// Returns an error if secret resolution or transport construction fails.
func New(cfg Config) (*Provider, error) {
	apiKey, err := provider.ResolveSecret(cfg.APIKey, cfg.SecretType)
	if err != nil {
		return nil, fmt.Errorf("amplitude: %w", err)
	}

	transport, err := provider.NewTransport(cfg.ProviderConfig)
	if err != nil {
		return nil, fmt.Errorf("amplitude: %w", err)
	}

	endpoint := cfg.Endpoint
	if endpoint == "" {
		endpoint = defaultEndpoint
	}

	p := &Provider{
		apiKey:      apiKey,
		endpoint:    endpoint,
		transport:   transport,
		rateLimiter: provider.NewRateLimiter(cfg.RateLimitConfig),
	}
	p.queue = provider.NewQueue(cfg.ProviderConfig, p.flushBatch)
	return p, nil
}

// Metadata returns the Amplitude provider's identity and capability set.
func (p *Provider) Metadata() provider.ProviderMetadata {
	return provider.ProviderMetadata{
		Name:    "amplitude",
		Version: version,
		Capabilities: provider.ProviderCapabilities{
			Track:    true,
			Identify: true,
			Group:    true,
			Page:     false,
			Alias:    true,
		},
	}
}

// Hooks returns no provider-level hooks.
func (p *Provider) Hooks() []hooks.Hook { return nil }

// Track enqueues a track event for batched delivery to Amplitude.
func (p *Provider) Track(ctx context.Context, msg provider.TrackMessage) error {
	return p.queue.Enqueue(ctx, provider.QueuedMessage{Op: "track", Track: &msg})
}

// Identify sends a $identify event synchronously. Identify bypasses the queue
// so that user-profile state arrives at Amplitude before any subsequent Track
// events that depend on it.
func (p *Provider) Identify(ctx context.Context, msg provider.IdentifyMessage) error {
	return p.sendBatch(ctx, []amplitudeEvent{mapIdentifyMessage(msg)})
}

// Group sends a $groupidentify event synchronously. Group bypasses the queue for
// the same reason as Identify — group membership must be established before
// subsequent Track events are attributed to the group.
func (p *Provider) Group(ctx context.Context, msg provider.GroupMessage) error {
	return p.sendBatch(ctx, []amplitudeEvent{mapGroupMessage(msg)})
}

// Page returns ErrUnsupportedOperation; Amplitude has no native page concept.
func (p *Provider) Page(_ context.Context, _ provider.PageMessage) error {
	return provider.ErrUnsupportedOperation
}

// Alias sends a $identify event synchronously, linking PreviousID to UserID.
// Alias bypasses the queue because identity merges must precede all subsequent
// events to avoid split-identity attribution.
func (p *Provider) Alias(ctx context.Context, msg provider.AliasMessage) error {
	return p.sendBatch(ctx, []amplitudeEvent{mapAliasMessage(msg)})
}

// Flush synchronously drains all buffered events to Amplitude.
func (p *Provider) Flush(ctx context.Context) error {
	return p.queue.Flush(ctx)
}

// Shutdown flushes remaining events and stops background processing.
func (p *Provider) Shutdown(ctx context.Context) error {
	return p.queue.Shutdown(ctx)
}

// flushBatch converts a batch of queued Track messages to amplitudeEvents and sends them.
// Only Track events are queued; Identify, Group, and Alias bypass the queue entirely.
func (p *Provider) flushBatch(ctx context.Context, batch []provider.QueuedMessage) error {
	events := make([]amplitudeEvent, 0, len(batch))
	for _, msg := range batch {
		if msg.Op == "track" {
			events = append(events, mapTrackMessage(*msg.Track))
		}
	}
	if len(events) == 0 {
		return nil
	}
	return p.sendBatch(ctx, events)
}

// sendBatch encodes events and POSTs them to the Amplitude batch endpoint,
// applying the configured rate limit before each HTTP request.
func (p *Provider) sendBatch(ctx context.Context, events []amplitudeEvent) error {
	if err := p.rateLimiter.Wait(ctx); err != nil {
		return err
	}

	payload := amplitudeBatchRequest{
		APIKey: p.apiKey,
		Events: events,
	}
	body, err := json.Marshal(payload)
	if err != nil {
		return fmt.Errorf("amplitude: marshal payload: %w", err)
	}

	effectiveURL, err := p.transport.RewriteURL(p.endpoint)
	if err != nil {
		return fmt.Errorf("amplitude: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, effectiveURL, bytes.NewReader(body))
	if err != nil {
		return fmt.Errorf("amplitude: build request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")

	resp, err := p.transport.Do(ctx, req, body)
	if err != nil {
		return fmt.Errorf("amplitude: HTTP request: %w", err)
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("amplitude: unexpected status %d", resp.StatusCode)
	}
	return nil
}

File: sdk/typescript/packages/provider-amplitude/src/provider.ts
import { EventQueue, ErrUnsupportedOperation } from '@dejanradmanovic/event-spec-api';
import type {
  Provider,
  ProviderMetadata,
  Hook,
  TrackMessage,
  IdentifyMessage,
  GroupMessage,
  PageMessage,
  AliasMessage,
} from '@dejanradmanovic/event-spec-api';
import type { AmplitudeConfig } from './config';
import { DEFAULT_ENDPOINT } from './config';
import { mapTrackMessage, mapIdentifyMessage, mapGroupMessage, mapAliasMessage } from './mapper';
import type { AmplitudeEvent } from './mapper';

// AmplitudeProvider sends analytics events to Amplitude via the batch HTTP API.
// Capabilities: Track ✅, Identify ✅, Group ✅, Page ❌ (unsupported), Alias ✅.
export class AmplitudeProvider implements Provider {
  private readonly apiKey: string;
  private readonly endpoint: string;
  private readonly maxRetries: number;
  private readonly initialBackoffMs: number;
  private readonly maxBackoffMs: number;
  private readonly retryMultiplier: number;
  private readonly jitter: boolean;
  private readonly rateLimitIntervalMs: number;
  private nextSendTime = 0;
  private readonly queue: EventQueue<TrackMessage>;

  constructor(config: AmplitudeConfig) {
    this.apiKey = config.apiKey;
    this.endpoint = resolveEndpoint(config);
    this.maxRetries = config.maxRetries ?? 3;
    this.initialBackoffMs = config.initialBackoffMs ?? 100;
    this.maxBackoffMs = config.maxBackoffMs ?? 30_000;
    this.retryMultiplier = config.retryMultiplier ?? 2.0;
    this.jitter = config.jitter ?? true;
    this.rateLimitIntervalMs =
      config.requestsPerSecond && config.requestsPerSecond > 0
        ? Math.ceil(1000 / config.requestsPerSecond)
        : 0;
    this.queue = new EventQueue<TrackMessage>((items) => this.flushBatch(items), {
      batchSize: config.batchSize ?? 100,
      flushIntervalMs: config.flushIntervalMs ?? 10000,
      maxSize: config.maxQueueSize ?? 10000,
      overflowPolicy: config.overflowPolicy ?? 'drop_oldest',
    });
  }

  metadata(): ProviderMetadata {
    return {
      name: 'amplitude',
      version: '0.1.0',
      capabilities: { track: true, identify: true, group: true, page: false, alias: true },
    };
  }

  hooks(): Hook[] {
    return [];
  }

  // track enqueues a track event for batched delivery.
  async track(msg: TrackMessage): Promise<void> {
    this.queue.enqueue(msg);
  }

  // identify sends a $identify event synchronously so user-profile state arrives at
  // Amplitude before any subsequent Track events that depend on it.
  async identify(msg: IdentifyMessage): Promise<void> {
    await this.sendBatch([mapIdentifyMessage(msg)]);
  }

  // group sends a $groupidentify event synchronously for the same reason as identify.
  async group(msg: GroupMessage): Promise<void> {
    await this.sendBatch([mapGroupMessage(msg)]);
  }

  // page is unsupported — Amplitude has no native page concept.
  async page(_msg: PageMessage): Promise<void> {
    throw ErrUnsupportedOperation;
  }

  // alias sends a $identify event synchronously, linking previousId to userId.
  async alias(msg: AliasMessage): Promise<void> {
    await this.sendBatch([mapAliasMessage(msg)]);
  }

  async flush(): Promise<void> {
    await this.queue.flushAll();
  }

  async shutdown(): Promise<void> {
    await this.queue.shutdown();
  }

  private async flushBatch(messages: TrackMessage[]): Promise<void> {
    const events = messages.map(mapTrackMessage);
    await this.sendBatch(events);
  }

  private async sendBatch(events: AmplitudeEvent[]): Promise<void> {
    const body = JSON.stringify({ api_key: this.apiKey, events });

    // Rate limiting: reserve a send slot and delay until it opens.
    if (this.rateLimitIntervalMs > 0) {
      const now = Date.now();
      const waitUntil = Math.max(now, this.nextSendTime);
      this.nextSendTime = waitUntil + this.rateLimitIntervalMs;
      if (waitUntil > now) {
        await sleep(waitUntil - now);
      }
    }

    let lastError: Error | undefined;
    let backoff = this.initialBackoffMs;
    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      if (attempt > 0) {
        const delay = this.jitter ? backoff * (0.5 + Math.random() * 0.5) : backoff;
        await sleep(delay);
        backoff = Math.min(backoff * this.retryMultiplier, this.maxBackoffMs);
      }
      try {
        const resp = await fetch(this.endpoint, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json', Accept: 'application/json' },
          body,
        });
        if (!resp.ok) {
          throw new Error(`amplitude: unexpected status ${resp.status}`);
        }
        return;
      } catch (err) {
        lastError = err instanceof Error ? err : new Error(String(err));
        if (attempt === this.maxRetries) break;
      }
    }
    throw lastError ?? new Error('amplitude: send failed');
  }
}

// resolveEndpoint computes the effective HTTP endpoint from config, applying proxy rewriting.
function resolveEndpoint(config: AmplitudeConfig): string {
  const base = config.endpoint ?? DEFAULT_ENDPOINT;
  if (!config.proxyUrl) return base;
  switch (config.proxyMode) {
    case 'reverse_proxy': {
      try {
        const proxy = new URL(config.proxyUrl);
        const target = new URL(base);
        target.protocol = proxy.protocol;
        target.host = proxy.host; // includes port when non-default
        if (proxy.pathname && proxy.pathname !== '/') {
          const basePath = proxy.pathname.replace(/\/$/, '');
          const provPath = target.pathname.replace(/^\//, '');
          target.pathname = `${basePath}/${provPath}`;
        }
        return target.toString();
      } catch {
        return base;
      }
    }
    case 'custom':
      return config.proxyUrl;
    default:
      return base;
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
